File: MorySnip/Classes/Polar.vb
Public Class Polar
    Public Angle As Single
    Public Radius As Single

    Public Shared Function Convert(Source As Polar) As PointF
        Return New PointF(Source.Radius * Math.Cos(Source.Angle), Source.Radius * Math.Sin(Source.Angle))
    End Function

    Public Shared Function Convert(Source As Point) As Polar
        Return Convert(New PointF(Source.X, Source.Y))
    End Function

    Public Shared Function Convert(Source As PointF) As Polar
        Dim Angle As Single, Radius As Single

        Dim X As Single = Source.X, Y As Single = Source.Y


        If X > 0 And Y >= 0 Then
            Angle = Math.Atan(Y / X)
        ElseIf X > 0 And Y < 0 Then
            Angle = Math.Atan(Y / X) + 2 * Math.PI
        ElseIf X < 0 Then
            Angle = Math.Atan(Y / X) + Math.PI
        ElseIf X = 0 And Y > 0 Then
            Angle = Math.PI / 2
        ElseIf X = 0 And Y < 0 Then
            Angle = Math.PI * 3 / 2
        ElseIf X = 0 And Y = 0 Then
            Angle = 0
        Else
            Throw New Exception("The impossible is possible")
        End If

        Radius = (X ^ 2 + Y ^ 2) ^ 0.5

        Return New Polar(Angle, Radius)
    End Function

    Public Shared Widening Operator CType(Source As Point) As Polar
        Return Convert(Source)
    End Operator

    Public Shared Widening Operator CType(Source As PointF) As Polar
        Return Convert(Source)
    End Operator

    Public Shared Widening Operator CType(Source As Polar) As Point
        Dim Result As PointF = Convert(Source)
        Return New Point(Result.X, Result.Y)
    End Operator

    Public Shared Widening Operator CType(Source As Polar) As PointF
        Return Convert(Source)
    End Operator

    Public Shared Widening Operator CType(Source As Polar) As Size
        Dim Result As PointF = Convert(Source)
        Return New Size(Result.X, Result.Y)
    End Operator

    Public Shared Widening Operator CType(Source As Polar) As SizeF
        Dim Result As PointF = Convert(Source)
        Return New SizeF(Result.X, Result.Y)
    End Operator

    Public Sub New()

    End Sub

    Public Sub New(Angle As Single, Radius As Single)
        Me.Angle = Angle
        Me.Radius = Radius
    End Sub
End Class

File: MorySnip/Classes/Polar.cs
using System;
using System.Drawing;
using Microsoft.VisualBasic.CompilerServices;

namespace Morysoft.MorySnip.Classes;

public class Polar
{
    public float Radius { get; set; }

    public float Angle { get; set; }

    public static PointF Convert(Polar source) => new(Conversions.ToSingle(source.Radius * Math.Cos(source.Angle)), Conversions.ToSingle(source.Radius * Math.Sin(source.Angle)));

    public static Polar Convert(Point source) => Convert(new PointF(source.X, source.Y));

    public static Polar Convert(PointF source)
    {
        float angle;
        float radius;
        var x = source.X;
        var y = source.Y;

        if (x > 0 && y >= 0)
        {
            angle = Conversions.ToSingle(Math.Atan(y / x));
        }
        else if (x > 0 && y < 0)
        {
            angle = Conversions.ToSingle(Math.Atan(y / x) + 2 * Math.PI);
        }
        else if (x < 0)
        {
            angle = Conversions.ToSingle(Math.Atan(y / x) + Math.PI);
        }
        else if (x == 0 && y > 0)
        {
            angle = Conversions.ToSingle(Math.PI / 2);
        }
        else if (x == 0 && y < 0)
        {
            angle = Conversions.ToSingle(Math.PI * 3 / 2);
        }
        else if (x == 0 && y == 0)
        {
            angle = 0;
        }
        else
        {
            throw new Exception("The impossible is possible");
        }

        radius = Conversions.ToSingle(Math.Pow(Math.Pow(x, 2) + Math.Pow(y, 2), 0.5));

        return new Polar(angle, radius);
    }

    public static implicit operator Polar(Point Source) => Convert(Source);

    public static implicit operator Polar(PointF Source) => Convert(Source);

    public static implicit operator Point(Polar Source)
    {
        var result = Convert(Source);

        return new Point(Conversions.ToInteger(result.X), Conversions.ToInteger(result.Y));
    }

    public static implicit operator PointF(Polar Source) => Convert(Source);

    public static implicit operator Size(Polar Source)
    {
        var result = Convert(Source);

        return new Size(Conversions.ToInteger(result.X), Conversions.ToInteger(result.Y));
    }

    public static implicit operator SizeF(Polar Source)
    {
        var result = Convert(Source);

        return new SizeF(result.X, result.Y);
    }

    public static Polar operator +(Polar left, Polar right) => Add(left, right);

    public static Polar operator -(Polar left, Polar right) => Subtract(left, right);

    public Polar() { }

    public Polar(float angle, float radius)
    {
        this.Angle = angle;
        this.Radius = radius;
    }

    public static Polar Add(Polar left, Polar right) => (PointF)left + (SizeF)right;

    public static Polar Subtract(Polar left, Polar right) => (PointF)left - (SizeF)right;
}
